### PR TITLE
Ref followups

### PR DESCRIPTION
**Pull request by OrionReed**
**Breaking changes:**
- Escape character changed from `~` to `\` (`%5C` in URLs)
- Removed indexed text refs
- `AutomergeRefUrl` → `RefUrl`
- `MatchPattern` → `Pattern`  
- `cursor()` must be the last segment in a path (throws otherwise)

**Features:**
- `ref.change(value)` shorthand for primitives: `ref.change("dark")` instead of `ref.change(() => "dark")`
- `RefOfType<T>` type utility for typed ref parameters
- `ref()` factory now returns same instance for identical paths (stable references)
- New exports: `cursor(from, to?)`, `fromUrl`

**Fixes:**
- Match patterns with slashes now survive URI encoding (JSON is fully URL-encoded)
- Double URI encoding/decoding no longer corrupts ref URLs

### CODE REVIEW

File: packages/refs/src/guards.ts [hunk 1]
```diff
@@ -1,16 +1,37 @@
...
 }
 
-/** Plain object used for id patterns (not a Segment or array) */
-export function isIdPattern(obj: unknown): obj is IdPattern {
-  if (obj === null || typeof obj !== "object" || Array.isArray(obj)) {
+export function isCursorMarker(val: unknown): val is CursorMarker {
+  return isObject(val) && CURSOR_MARKER in val;
+}
+
+export function isMatchPattern(val: unknown): val is MatchPattern {
```
Comment (paulsonnentag) [inline packages/refs/src/guards.ts]: Can we just call it pattern? MatchPattern sounds redundant

File: packages/refs/src/parser.ts [hunk 2]
```diff
@@ -1,228 +1,202 @@
...
+ * | Type    | Format              | Example                    | Notes                           |
+ * |---------|---------------------|----------------------------|---------------------------------|
+ * | Key     | string              | `foo`, `my%2Fkey`          | Default, URL-encoded            |
+ * | Index   | `@` + number        | `@0`, `@42`                | Array index                     |
+ * | Match   | `{...}`             | `{"id":"alice"}`           | JSON object pattern             |
+ * | Cursors | `[start-end]`       | `[2@abc-5@def]`            | Cursor range                    |
+ * | Cursors | `[cursor]`          | `[2@abc]`                  | Collapsed (start === end)       |
+ *
+ * ## Escape Rule
+ * If a key starts with `@`, `{`, `[`, or `~`, prefix with `~`:
```
Comment (paulsonnentag) [inline packages/refs/src/parser.ts]: I don't like "~" as an escape character could we use "\\" instead it's a bit more awkward to use in strings but it's more familiar to understand if you are looking at it and don't know how ref urls work

File: packages/refs/src/parser.ts [hunk 3]
```diff
@@ -1,228 +1,202 @@
...
- * Parse a URL path string into segments.
- *
- * @example
- * parsePath("todos/0/title") → [{ kind: "key", key: "todos" }, ...]
- * parsePath("todos/:abc123") → [{ kind: "key", key: "todos" }, { kind: "stable_index", id: "abc123" }]
- * parsePath("note/0..10") → [..., { kind: "range", start: 0, end: 10 }]
+ * Codecs in priority order. First matching codec wins for parsing.
+ * Order: index → match → cursors → key (key is catch-all, must be last)
  */
+const SEGMENT_CODECS = [
```
Comment (paulsonnentag) [inline packages/refs/src/parser.ts:172]: I would prefer to just use function calls and make each segment codec just a function instead of having this abstraction of segment codecs.

Comment (paulsonnentag) [reply packages/refs/src/parser.ts:172]: Overall I like this pattern though of splitting the path and having little parsers for the different patterns

File: packages/refs/src/types.ts [hunk 4]
```diff
@@ -94,3 +116,92 @@ export type InferRefType<TDoc, TPath extends readonly any[]> = PathValue<
...
+export type InferRefTypeFromString<
+  TDoc,
+  P extends string,
+> = PathValueFromString<TDoc, SegmentsFromString<P>>;
+
+/**
+ * Branded type for Automerge ref URLs.
+ * A string in the format: `automerge:documentId/path#heads`
+ */
+export type AutomergeRefUrl = string & { readonly __brand: "AutomergeRefUrl" };
```
Comment (paulsonnentag) [inline packages/refs/src/types.ts]: Could we simplify the name and just call them refUrls? I think the long term goal should be to eventually fold RefUrls into AutomergeUrls. Like an AutomergeUrl could point to either a whole document or a part of it

File: packages/refs/src/test/ref.test.ts [hunk 5]
```diff
@@ -119,36 +115,37 @@ describe("Ref", () => {
       });
 
       const themeRef = new Ref(handle, ["user", "settings", "theme"]);
-      themeRef.change((theme) => "dark");
+      themeRef.change(() => "dark");
```
Comment (paulsonnentag) [inline packages/refs/src/test/ref.test.ts:118]: could we have shorthand that you could do themeRef.change("dark") for primitive values?

File: packages/refs/src/test/edge-cases.test.ts [hunk 6]
```diff
@@ -0,0 +1,683 @@
...
+    });
+
+    it("should ignore segments after cursor() (terminates path)", () => {
+      handle.change((d) => {
+        d.text = "Hello World";
+      });
+
+      // cursor() terminates path normalization, "invalid" is ignored
+      expect(() => {
+        new Ref(handle, ["text", cursor(0, 5), "invalid"]);
```
Comment (paulsonnentag) [inline packages/refs/src/test/edge-cases.test.ts:82]: I think this should an error

**Review by paulsonnentag — APPROVED** [review]
Looks good to me. I left some comments

File: packages/refs/src/parser.ts (hunk 3 above)
Comment (OrionReed) [reply packages/refs/src/parser.ts:172]: I did have that before, but when making changes I managed to introduce bugs a couple of times from having the source of truth spread out. I also don't like the introduced abstraction but in this case I think its worth it for the peace of mind (very open to changing this down the line, but will leave as-is for now)

Comment (paulsonnentag) [reply packages/refs/src/parser.ts:172]: That makes sense

**Comment by OrionReed** [comment]
TODO: add `addComment(thread: RefOfType<Todo>)` RefOfType ts util

File: packages/refs/src/parser.ts [hunk 7]
```diff
@@ -1,232 +1,242 @@
...
-
-/**
- * Parse a URL path string into segments.
+ * # Path Segment Encoding Scheme
+ *
+ * | Type    | Format              | Example                    | Notes                           |
+ * |---------|---------------------|----------------------------|---------------------------------|
+ * | Key     | string              | `foo`, `my%2Fkey`          | Default, URL-encoded            |
+ * | Index   | `@` + number        | `@0`, `@42`                | Array index                     |
+ * | Match   | `{...}`             | `{"id":"alice"}`           | JSON object pattern (URL-encoded) |
```
Comment (paulsonnentag) [inline packages/refs/src/parser.ts:12]: I think this should be called pattern

**Review by paulsonnentag — APPROVED** [review]
Overall this looks good to me. I've already merged this into my annoation branch and it feels good to use. 

A couple smaller notes:

- I think loadRef should take a repo and not a doc handle so you can easily load refs where you don't know which document they point to
- I need `Ref` in annotations we should export just the type of it
- `RefOfType` requires type casts some times to use it as a ref again. Here is an example:
```ts
...
  onChildrenOf(
    ref: RefOfType<string | Array<unknown>>
  ): FilteredAnnotationView {
    return new FilteredAnnotationView(this, (otherRef, _) =>
      otherRef.isChildOf(ref as Ref) // need to do a type cast here
    );
  }
...
```
- I think we need to pick another package name so we can publish it `@inkandswitch/refs` maybe?

A bigger question is how we handle multiple instances of a library. The way that refs is implemented right now means that we need to share a single instance of the refs library across tools because it use symbol and a local ref cache.

I would prefer if we build our libraries in a way that we don't have to externalize them to make them less special and more like something anyone could build.
